### DISCUSSION TITLE
Issues with nested contains_eager and multiple aliases

### DISCUSSION THREAD

**Discussion by sonthonaxrk** [sqlalchemy/sqlalchemy#7764] [Usage Questions]
Hey all :)

I'm not sure if this is a bug, or just an API I'm failing to understand; but I'm trying to model the following data model, where I have multiple foreign keys to the same table.  Like so:

![image](https://user-images.githubusercontent.com/8541009/155508957-63597316-61f9-4a20-b414-b7eda78644a0.png)

Generating the SQL for this is trivial, however the ORM seems to be loading the wrong objects.  I've written a number of tests all trying to do essentially the same thing: that is, to load _Order -> Customer -> Account_ with `contains_eager`.   As far as I can tell, I can only get `contains_eager` to work one layer deep.

I dug into orm/loading.py and found that the ColumnAdapter used for _Order -> Customer -> Account_, is actually the adapter used for _Order -> Account_.   I think this is a bug.

If you run the below with `pytest` you can see the various attempts I've made to sidestep this issue.  All of the tests fail.

(Just rewriting test, didn't demonstrate issue)

**Comment by zzzeek** [comment]
make sure to add a comment here so i know to come back.  your first example in my email had a lot of things that didn't look right.

**Reply by sonthonaxrk** [reply]
Sorry about that.  I'm just a bit at a loss at how to reproduce it outside the system I'm running it in.  There's definitely a weird referencing issue going on.

**Reply by zzzeek** [reply]
put up the code that you think is what should produce the problem, but doesnt, because the code I saw had a lot of issues with it and it might be data-related that you actually see something going wrong.

**Reply by zzzeek** [reply]
for example things like "Customer.account.of_type(Account)" makes no sense, it looked like you had aliases going on so you'd want to use aliased() and "Customer.account.of_type(aliased_account)"  .

**Reply by sonthonaxrk** [reply]
OMG, I think I've found the issue.  Sorry for the incorrect examples, I was just trying to reproduce in the basic case, and out of pure frustration I was emulating but not reproducing the error.

Basically I found a stop gap solution that involved adding the entities to the query, but disregarding them, instead continuing to load them using contains eager.  Semantically this shouldn't have made a difference, but it changed how the ColumnAdapters worked in practice. 

I was also using a lot of aliases, because I was joining the same table multiple times on different FKs.  However this has a weird interaction (where I was aliasing an alias) with weakrefs in the AliasedInsp class.  I removed the weakref in the `AliasedInsp.__init__` method just to confirm this was the problem and I no longer had to add the entities manually to the base query.

Oh god this was annoying to find out.

Totally appreciate this is mega fringe issue that probably only I've experienced.  But thanks for your help regardless - I appreciate the time you spend trying to help everyone who asks a question here :)

**Reply by zzzeek** [reply]
no worries, this is why it's called......  **alchemy** !!
